File: scripts/generate_events.rb
require 'open-uri'
require 'json'
require 'date'
require 'time'

ICS_URL = "https://calendar.google.com/calendar/ical/a9990f54cc62171dfef7a892c552c8964dc1a18735784e6d8c36803ce3cffb9b%40group.calendar.google.com/public/basic.ics"

ics = URI.open(ICS_URL).read

events = []

ics.split("BEGIN:VEVENT")[1..]&.each do |block|
  body = block.split("END:VEVENT")[0]

  summary = body[/SUMMARY:(.+)/,1]&.strip
  next unless summary

  # DTSTART handling
  start_raw =
    if body =~ /DTSTART;VALUE=DATE:(\d{8})/
      d = Date.strptime($1, "%Y%m%d")
      Time.utc(d.year, d.month, d.day, 12, 0, 0) # normalize to noon UTC
    else
      raw = body[/DTSTART(?:;[^:]+)?:([0-9T]+Z?)/,1]
      begin
        Time.parse(raw).utc
      rescue
        nil
      end
    end

  next unless start_raw

  events << {
    summary: summary,
    startRaw: start_raw.iso8601,
    allDay: (body =~ /DTSTART;VALUE=DATE:/ ? true : false)
  }
end

# sort by actual date only (ignore google push delay)
events.sort_by! { |e| Time.parse(e[:startRaw]) }

File.write("assets/events.json", JSON.pretty_generate({
  generatedAt: Time.now.utc.iso8601,
  events: events
}))

puts "Generated assets/events.json with #{events.size} events"
puts File.read("assets/events.json")  # SHOW EXACT JSON IN WORKFLOW LOG

File: scripts/generate-events.mjs
const ICS_URL = process.env.ICS_URL;

const text = await fetch(ICS_URL).then(r => {
  if (!r.ok) throw new Error("Failed to fetch ICS: " + r.status);
  return r.text();
});

// Split into VEVENT blocks
const blocks = text.split("BEGIN:VEVENT").slice(1);

// Minimal parser: pull out DTSTART and SUMMARY (and note all-day)
const events = [];
for (const ch of blocks) {
  const endIdx = ch.indexOf("END:VEVENT");
  const body = endIdx >= 0 ? ch.slice(0, endIdx) : ch;

  // Prefer explicit all-day first
  let allDay = false;
  let startRaw = null;

  const mAll = body.match(/DTSTART;VALUE=DATE:(\d{8})/);
  if (mAll) {
    allDay = true;
    startRaw = mAll[1]; // YYYYMMDD
  } else {
    // Any DTSTART variant, grab value after colon
    const mAny = body.match(/DTSTART(?:;[^:]+)?:([0-9T]+Z?)/);
    if (mAny) startRaw = mAny[1]; // e.g., 20251106T130000Z or 20251106T130000
  }

  const mSum = body.match(/SUMMARY:(.+)/);

  if (startRaw && mSum) {
    events.push({
      summary: mSum[1].trim(),
      startRaw,
      allDay
    });
  }
}

// Write full list; footer will sort/filter client-side in ET.
const fs = await import('node:fs/promises');
await fs.writeFile('assets/events.json', JSON.stringify({ generatedAt: new Date().toISOString(), events }, null, 2));
console.log("Wrote assets/events.json with", events.length, "events");
